File: script-files/measures/outputsContinuous/t_student.js
jStat = require("../node_modules/jstat");

dists = {
    t_student: {
        measurements: {
            nonCentralMoments: function (k, df) {
                return undefined;
            },
            centralMoments: function (k, df) {
                return undefined;
            },
            stats: {
                mean: function (df) {
                    return 0;
                },
                variance: function (df) {
                    return df / (df - 2);
                },
                standardDeviation: function (df) {
                    return Math.sqrt(this.variance(df));
                },
                skewness: function (df) {
                    return 0;
                },
                kurtosis: function (df) {
                    return 6 / (df - 4) + 3;
                },
                median: function (df) {
                    return dists.t_student.measurements.ppf(0.5, df);
                },
                mode: function (df) {
                    return 0;
                },
            },
        }
    }
}
console.log(dists.t_student.measurements.stats.mean(10))
console.log(dists.t_student.measurements.stats.variance(10))
console.log(dists.t_student.measurements.stats.standardDeviation(10))
console.log(dists.t_student.measurements.stats.skewness(10))
console.log(dists.t_student.measurements.stats.kurtosis(10))
// console.log(dists.t_student.measurements.stats.median(10))
console.log(dists.t_student.measurements.stats.mode(10))

// mean_value: 0
// variance_value: 1.25
// standard_deviation_value: 1.0772477334018427
// skewness_value: 0
// kurtosis_value: 4
// median_value: 0
// mode_value: 0

File: continuous/distributions/t_student.py
import scipy.stats
import scipy.special
import numpy


class T_STUDENT:
    """
    T distribution
    https://en.wikipedia.org/wiki/Student%27s_t - distribution
    Hand - book on  STATISTICAL  DISTRIBUTIONS  for  experimentalists (pag.143) ...  Christian Walck
    """

    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)
        self.df = self.parameters["df"]

    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        # result = scipy.stats.t.cdf(x, self.df)
        result = scipy.special.betainc(self.df / 2, self.df / 2, (x + numpy.sqrt(x * x + self.df)) / (2 * numpy.sqrt(x * x + self.df)))
        return result

    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        # result = scipy.stats.t.pdf(x, self.df)
        result = (1 / (numpy.sqrt(self.df) * scipy.special.beta(0.5, self.df / 2))) * (1 + x * x / self.df) ** (-(self.df + 1) / 2)
        return result

    def ppf(self, u):
        # result = scipy.stats.t.ppf(u, self.df)
        if u >= 0.5:
            result = numpy.sqrt(self.df * (1 - scipy.special.betaincinv(self.df / 2, 0.5, 2 * min(u, 1 - u))) / scipy.special.betaincinv(self.df / 2, 0.5, 2 * min(u, 1 - u)))
            return result
        else:
            result = -numpy.sqrt(self.df * (1 - scipy.special.betaincinv(self.df / 2, 0.5, 2 * min(u, 1 - u))) / scipy.special.betaincinv(self.df / 2, 0.5, 2 * min(u, 1 - u)))
            return result

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.df > 0
        return v1

    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by solving the equations of the measures expected
        for this distribution.The number of equations to consider is equal to the number
        of parameters.

        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"alpha": * , "beta": * , "min": * , "max": * }
        """

        df = 2 * measurements.variance / (measurements.variance - 1)

        parameters = {"df": df}

        return parameters


if __name__ == "__main__":
    ## Import function to get measurements
    import sys
    import numpy

    sys.path.append("../measurements")
    from measurements_continuous import MEASUREMENTS_CONTINUOUS

    ## Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data

    ## Distribution class
    path = "../data/data_t_student.txt"
    data = get_data(path)
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = T_STUDENT(measurements)

    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.cdf(numpy.array([measurements.mean, measurements.mean])))
    print(distribution.pdf(measurements.mean))
    print(distribution.pdf(numpy.array([measurements.mean, measurements.mean])))
    print(distribution.cdf(-3))
    print(distribution.pdf(-3))

    print(distribution.ppf(0.007627123509096105))
    print(distribution.ppf(0.9))

    print(scipy.stats.t.fit((measurements.data)))
